Add support for clip skip

Efreak opened the pull request:
<!-- Replace [Title] with a short summary of what this pull request is -->
# Add support for Clip Skip and extra lora fields

## Detailed Description

This PR:
1. Adds support for clip_skip
  - Applies clip_skip set in styles when using /generate
  - Applies clip_skip set in styles when using /advanced_generate
  - Adds a subcommand to set clip_skip when using /advanced_generate
  - Adds default settings for clip_skip
2. Adds support for the extra lora fields in styles in /generate
3. When user doesn't set loras in /advanced_generate, use style loras

## Why this is needed
1. clip skip is *required* for pony models, they return blurry fog without clip skip set to >1. The latest version of ICBINP is also superior with clip skip.
2. there's currently no bot support for lora versions, which are used in multiple styles
3. advanced_generate has no option to use multiple loras as required by some styles, so it's currently impossible to use them properly and tweak other settings.


## Other Info
I've used up the last available subcommand in /advanced_generate. This can be mitigated:
- One can be freed up by using logic regarding denoising and source image (if denoising is set, enable hires-fix. provide an autocomplete response of True and False that set denoising value at 65 and 0 internally, and disable hiresfix if denoise is set to 0. (65 is the default)
- One can be freed up by merging postprocessors, which will also allow adding all the possible postprocessors (rembg, choice of 5 upscalers, codeformers, gfpgan). For initial autocomplete, show all options; if there's a valid option ending with a comma, autocomplete can provide the remaining valid options (you can only have one upscaler, but it can be mixed with both face fixers and rembg). Since order of postprocessing is not customizable, number of possible responses can be reduced by not allowing arbitrary order.

Efreak committed:
add support for clip skip in styles

Efreak committed:
add support for more fields from styles in generate command

Efreak committed:
add clip_skip subcommand for advanced_generate

Efreak committed:
use style loras on /advanced_generate if none specified

ZeldaFan0225 commented:
Has this code been tested?

ZeldaFan0225 commented on src/commands/advanced_generate.ts:
```diff
@@ -287,6 +297,7 @@ export default class extends Command {
         const ti_raw = ctx.interaction.options.getString("textual_inversion") ?? ctx.client.config.advanced_generate.default?.tis
         const hires_fix = ctx.interaction.options.getBoolean("hires_fix") ?? ctx.client.config.advanced_generate.default?.hires_fix ?? false
         const qr_code_url = ctx.interaction.options.getString("qr_code_url")
```
Don't hard code defaults, if none are given in the config the horde default should be applied

Efreak replied:
Hires-fix does default to false (it's not even supported on SDXL payloads), but I'll change it to undefined instead.

ZeldaFan0225 commented on src/commands/generate.ts:
```diff
@@ -142,6 +142,7 @@ export default class extends Command {
 
         const style = ctx.client.getHordeStyle(style_raw)
 
```
Same thing as the previous comment

Efreak replied:
Assuming you mean clip skip at line 145 (not shown for some reason), 1 is the default as mentioned in the comment above, but I'll set it to undefined here as well.

ZeldaFan0225 replied:
Same thing as above, remove the ?? undefined

Efreak committed:
no hardcoded defaults

Efreak committed:
make clip skip configurable (in bot configuration) for /generate

Some UIs (including forge) default clip skip to 2, so allow this to be a
default for the bot as well.

I've been told that the fact SDXL itself works at clip skip 1 is
accidental, and it's intended to be used at 2 (I haven't verified this, and
don't know where I would verify this)

ZeldaFan0225 commented on src/commands/advanced_generate.ts:
```diff
@@ -295,9 +295,9 @@ export default class extends Command {
         const lora_id = ctx.interaction.options.getString("lora")
         const ti_raw = ctx.interaction.options.getString("textual_inversion") ?? ctx.client.config.advanced_generate.default?.tis
-        const hires_fix = ctx.interaction.options.getBoolean("hires_fix") ?? ctx.client.config.advanced_generate.default?.hires_fix ?? false
+        const hires_fix = ctx.interaction.options.getBoolean("hires_fix") ?? ctx.client.config.advanced_generate.default?.hires_fix ?? undefined
```
Remove the whole ?? undefined because the default from config already is either a value or undefined

ZeldaFan0225 commented on src/commands/advanced_generate.ts:
```diff
@@ -295,9 +295,9 @@ export default class extends Command {
+        const hires_fix = ctx.interaction.options.getBoolean("hires_fix") ?? ctx.client.config.advanced_generate.default?.hires_fix ?? undefined
         const qr_code_url = ctx.interaction.options.getString("qr_code_url")
-        const clipskip = ctx.interaction.options.getInteger("clip_skip") ?? style?.clip_skip ?? ctx.client.config.advanced_generate?.default?.clip_skip ?? 1
+        const clipskip = ctx.interaction.options.getInteger("clip_skip") ?? style?.clip_skip ?? ctx.client.config.advanced_generate?.default?.clip_skip ?? undefined
```
Also get rid of the ?? undefined here

Efreak committed:
remove unnecessary undefined statements

the previous statements already return undefined if necessary, this isn't `||` (I keep forgetting the difference)
